### Library

In [1]:
import pandas as pd

Imported libraries what will use.

### Analyze Datas

In [5]:
dataframe = pd.read_csv("flo100k.csv", sep="|")
dataframe.head()

master_id order_channel platform_type  \
0  b3ace094-a17f-11e9-a2fc-000d3a38a36f       Offline       Offline   
1  c57d7c4c-a950-11e9-a2fc-000d3a38a36f       Offline   OmniChannel   
2  602897a6-cdac-11ea-b31f-000d3a38a36f       Offline       Offline   
3  388e4c4e-af86-11e9-a2fc-000d3a38a36f        Mobile        Online   
4  80664354-adf0-11eb-8f64-000d3a299ebf       Desktop        Online   

  last_order_channel               first_order_date  \
0            Offline  2019-02-23T12:59:17.000+03:00   
1            Offline  2019-12-01T16:48:09.000+03:00   
2            Offline  2020-07-24T15:49:47.000+03:00   
3             Mobile  2018-12-31T07:22:07.000+03:00   
4            Desktop  2021-05-05T21:07:02.000+03:00   

                 last_order_date         last_order_date_online  \
0  2019-02-23T12:59:17.000+03:00                            NaN   
1  2019-12-01T16:48:09.000+03:00                            NaN   
2  2020-07-24T15:49:47.000+03:00                            NaN   
3  2018-12-31T07:22:07.000+03:00  2018-12-31T07:22:07.000+03:00   
4  2021-05-05T22:39:36.000+03:00  2021-05-05T22:39:36.000+03:00   

         last_order_date_offline  order_num_total_ever_online  \
0  2019-02-23T12:59:17.000+03:00                          NaN   
1  2019-12-01T16:48:09.000+03:00                          NaN   
2  2020-07-24T15:49:47.000+03:00                          NaN   
3                            NaN                          1.0   
4                            NaN                          2.0   

   order_num_total_ever_offline  customer_value_total_ever_offline  \
0                           1.0                             212.98   
1                           1.0                             199.98   
2                           1.0                             140.49   
3                           NaN                               0.00   
4                           NaN                               0.00   

   customer_value_total_ever_online interested_in_categories_12  \
0                              0.00                         NaN   
1                              0.00                         NaN   
2                              0.00                     [ERKEK]   
3                            174.99                         NaN   
4                            283.95                          []   

  online_product_group_amount_top_name_12 offline_product_group_name_12  \
0                                     NaN                           NaN   
1                                     NaN                           NaN   
2                                     NaN                         ERKEK   
3                                     NaN                           NaN   
4                                     NaN                           NaN   

  last_order_date_new store_type  
0          2019-02-23          A  
1          2019-12-01          A  
2          2020-07-24          A  
3          2018-12-31          A  
4          2021-05-05          A

In [6]:
dataframe.head().to_excel("output.xlsx")

There is too feature when see datas. When analyze to features, company; 
* What do customers prefer platform type?
* Which channel do customers use on the platform?
* What is the customer's budget for shopping?
* Which categories do customers shop?
finding answers to questions.

#### Features

| Features Name | Data Type | Description |
| --- | --- | --- |
| **master_id** | string | *Customer id* |
| **order_channel** | string | *Which channel used for shopping (Android, ios, Desktop, Mobile)* |
| **platform_type** | string | *Shopped platform type. (Online, Offline, Omnichannel)* |
| **last_order_channel** | string | *last shopped channel* |
| **first_order_date** | date | *First date of shopped* |
| **last_order_date** | date | *Last date of shopped* |
| **last_order_date_online** | date | *Last date of shopped on online platform* |
| **last_order_date_offline** | date | *Last date of shopped on offline platform* |
| **order_num_total_ever_online** | int |*Total shopped count on online platform* |
| **order_num_total_ever_offline** | int | *Total shopped count on offline platform* |
| **customer_value_total_ever_offline** | float | *Total paid for shopped on offline platform* |
| **customer_value_total_ever_online** | float | *Total paid for shopped on online platform* |
| **interested_in_categories_12** | string | *Shooped which in categories to last 12 months* |
| **online_product_group_amount_top_name_12** | string | *Which product group in shopped to last 12 months* |
| **offline_product_group_name_12** | string | *Shooped which in categories to last 12 months on offline platform* |
| **last_order_date_new** | date | *Last order date* |
| **store_stype** | string | *It contains 3 different companies. If customer shops A and B, It writes A,B.* |

### Dataframe Schemas for PostgreSQL & Apache Hive

![Alt text](schemas.jpg) 

For these schemas, see to features names and their data types. Data types can be change what sql uses. They are for postgresql and apache hive. Seemed the dataframe structure. From now on, continue with postgresql schema.

For application what we can do;
1. Create table for PostgreSQL schema.
2. Importing csv file datas to table.
3. Transfer table PostgreSQL to Apache HDFS with Apache Sqoop.
4. Transfer table PostgreSQL to Apache Hive with Apache Sqoop.
5. Create another Apache Hive table as .orc formatted.
6. Analyse data on Apache Hive.

### Apache: HDFS, Hive, Sqoop and PostgreSQL Application

In this chapter, see how to communicate apache tools and postgresql each other. Firstly, create table for dataframe. After, importing csv file to postgresql. Then transfer all table with datas to apache hdfs and hive with apache sqoop. This transferred table is not good efficiency therefore convert table to .orc formatted table. Finally, few data analyses on apache hive.

#### Creating Part

<span style="color:red">create table if not exists </span> customers ( <br />
&nbsp;&nbsp;&nbsp;&nbsp;	master_id <span style="color:blue">varchar(100)</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	order_channel <span style="color:blue">varchar(100)</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	platform_type <span style="color:blue">varchar(25)</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	last_order_channel <span style="color:blue">varchar(25)</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	first_order_date <span style="color:blue">varchar(75)</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	last_order_date <span style="color:blue">varchar(75)</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	last_order_date_online <span style="color:blue">varchar(75)</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	last_order_date_offline <span style="color:blue">varchar(75)</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	order_num_total_ever_online <span style="color:red">float</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	order_num_total_ever_offline <span style="color:red">float</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	customer_value_total_ever_offline <span style="color:red">float</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	customer_value_total_ever_online <span style="color:red">float</span>,  <br />
&nbsp;&nbsp;&nbsp;&nbsp;	interested_in_categories_12 <span style="color:blue">varchar(75)</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	online_product_group_amount_top_name_12 <span style="color:blue">varchar(75)</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	offline_product_group_name_12 <span style="color:blue">varchar(75)</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	last_order_date_new <span style="color:blue">varchar(75)</span>, <br />
&nbsp;&nbsp;&nbsp;&nbsp;	store_type <span style="color:blue">varchar(50)</span> <br />
);

On the above, create table without data. That's means setting the column name and its types and it is ready for upcoming datas.

#### Importing Part

```
psql -d traindb -c "\copy customers from 'flo100k.csv' DELIMITERS '|' CSV HEADER;"
```

It is console command. Postgre uses copy command for importing csv files to table. In the command firstly call to postgresql and then selecting which database and sql command.

#### Transportation Part

`
sqoop import --connect jdbc:postgresql://localhost:5432/traindb --driver org.postgresql.Driver --username train --password-file file:///home/train/.sqoop.password --table customers --split-by master_id --hive-import --create-hive-table --hive-table test1.customer --target-dir /tmp/c
`

On the above command, sqoop connected to postgresql server. After that give driver, username, password, which table in postgresql, which separator column and hive commands. For hive commands, it is new table so that used to create hive table. Finally create temp file for transportation.

`SELECT * FROM customer limit 5;`

master_id                           |order_channel|platform_type|last_order_channel|first_order_date             |last_order_date              |last_order_date_online       |last_order_date_offline      |order_num_total_ever_online|order_num_total_ever_offline|customer_value_total_ever_offline|customer_value_total_ever_online|interested_in_categories_12|online_product_group_amount_top_name_12|offline_product_group_name_12|last_order_date_new|store_type|
------------------------------------|-------------|-------------|------------------|-----------------------------|-----------------------------|-----------------------------|-----------------------------|---------------------------|----------------------------|---------------------------------|--------------------------------|---------------------------|---------------------------------------|-----------------------------|-------------------|----------|
b3ace094-a17f-11e9-a2fc-000d3a38a36f|Offline      |Offline      |Offline           |2019-02-23T12:59:17.000+03:00|2019-02-23T12:59:17.000+03:00|null                         |2019-02-23T12:59:17.000+03:00|                           |                         1.0|                           212.98|                             0.0|null                       |null                                   |null                         |2019-02-23         |A         |
c57d7c4c-a950-11e9-a2fc-000d3a38a36f|Offline      |OmniChannel  |Offline           |2019-12-01T16:48:09.000+03:00|2019-12-01T16:48:09.000+03:00|null                         |2019-12-01T16:48:09.000+03:00|                           |                         1.0|                           199.98|                             0.0|null                       |null                                   |null                         |2019-12-01         |A         |
602897a6-cdac-11ea-b31f-000d3a38a36f|Offline      |Offline      |Offline           |2020-07-24T15:49:47.000+03:00|2020-07-24T15:49:47.000+03:00|null                         |2020-07-24T15:49:47.000+03:00|                           |                         1.0|                           140.49|                             0.0|[ERKEK]                    |null                                   |ERKEK                        |2020-07-24         |A         |
388e4c4e-af86-11e9-a2fc-000d3a38a36f|Mobile       |Online       |Mobile            |2018-12-31T07:22:07.000+03:00|2018-12-31T07:22:07.000+03:00|2018-12-31T07:22:07.000+03:00|null                         |                        1.0|                            |                              0.0|                          174.99|null                       |null                                   |null                         |2018-12-31         |A         |
80664354-adf0-11eb-8f64-000d3a299ebf|Desktop      |Online       |Desktop           |2021-05-05T21:07:02.000+03:00|2021-05-05T22:39:36.000+03:00|2021-05-05T22:39:36.000+03:00|null                         |                        2.0|                            |                              0.0|                          283.95|[]                         |null                                   |null                         |2021-05-05         |A         |

As seen, transportation was successed.

#### Optimization Part

In here, optimize to apache hive. That's mean apache hive gives the high performance for data format. The best performance is .orc formatted table. Given command below create table from created table with file format.

`CREATE TABLE customer_orc stored as orc as SELECT * FROM test1.customer;`

After the conversion, table storage decreases fastly.

`hdfs dfs -du -h /user/hive/warehouse/test1.db | grep customer`

| Storage | File |
| --- | --- |
| 20.1 M | /user/hive/warehouse/test1.db/customer |
| 4.3 M | /user/hive/warehouse/test1.db/customer_orc |

#### Data Analyses Part

It is basic queries but important. First is given us to total count for the store type, second is given us to total count for the order channel.

1)

`SELECT store_type , COUNT(*) as total_count FROM customer_orc 
GROUP BY store_type
ORDER BY total_count DESC;`

store_type|total_count|
----------|-----------|
A         |      89225|
A,B       |       8498|
B         |       1491|
A,C       |        702|
A,B,C     |         75|
B,C       |         10|

2.

`SELECT order_channel, COUNT(*) as total_count FROM customer_orc
GROUP BY order_channel 
ORDER BY total_count DESC;`

order_channel|total_count|
-------------|-----------|
Offline      |      70785|
Android App  |      11989|
Mobile       |       8512|
Desktop      |       4751|
Ios App      |       3964|

### Conclusion

Interested in Apache HDFS, Hive, Sqoop. It is all how to use them each other. Project is for being ready for real project. Project is maden for who want to be data engineer.